### DISCUSSION TITLE
How should I login inside app?

### DISCUSSION THREAD

**Discussion by iosYash** [twilio/twilio-conversations-demo-ios-swift#14] [General]
What should be username and password for login?

**Comment by Gray-Wind** [comment]
Hello, the demo application fetches a token from a given backend service. It is not safe to store any sensitive information (like auth and secret keys) inside the application.

You can set up your own token provider with a basic authorization. In the app settings (iOS Settings → Conversions App → Token Service URL) you can add the URL. (the option in the settings bundle: https://github.com/twilio/twilio-conversations-demo-ios-swift/blob/main/ConversationsApp/Settings.bundle/Root.plist#L15)

The url with auth details is being constructed here: https://github.com/twilio/twilio-conversations-demo-ios-swift/blob/main/ConversationsApp/ConversationsClient/TokenWrapperImpl.swift#L17-L38
Feel free to modify the code to your needs.

**Reply by iosYash** [reply]
Where should I get identity, username & password?

**Reply by Gray-Wind** [reply]
I suggest to start with this doc: https://www.twilio.com/docs/conversations/create-tokens
It covers the identity part, which is being used by the conversation service. 

Setting a username and a password for the token generation service is your own responsibility in order to protect your Twilio secrets.

**Reply by iosYash** [reply]
Hello @Gray-Wind,

I am able to create accesstoken but inside the demo app when going to create conversation client it gives me an error "UNAUTHORISED". 

here is the code where I am getting error

```
func create(login: String, password: String, tokenWrapper: TokenWrapper.Type, delegate: TwilioConversationsClientDelegate, completion: @escaping (LoginResult) -> Void) {
        self.tokenWrapper = tokenWrapper
        tokenWrapper.getConversationsAccessToken(identity: login, password: password) { (result) in
            switch result {
            case .failure(let error):
                completion(.failure(error))
            case .success(let conversationsToken):
                // Create conversations client with token
                TwilioConversationsClient.setLogLevel(.silent)

                let properties = TwilioConversationsClientProperties()
                properties.dispatchQueue = DispatchQueue(label: "TwilioConversationsDispatchQueue")
                TwilioConversationsClient.conversationsClient(withToken: conversationsToken,
                                                              properties: properties,
                                                              delegate: delegate) { [weak self] result, client in
                    DispatchQueue.main.async {
                        if result.isSuccessful, let client = client  {
                            self?.conversationsClient = client
                            completion(.success)
                        } else {
                            completion(.failure(result.error!))
                        }
                    }
                }
            }
        }
    }
```


I have double check I have granted permissions for SERVICE_ID in serverless function. But don't know what exactly the issue is.
